docs/source/development/AlgorithmMPISupport.rst

```diff
@@ -0,0 +1,464 @@
+
+
+Master-only execution
+---------------------
+
+Master-only execution is handled by ``Algorithm::execMasterOnly()``.
+By default this simply calls ``Algorithm::exec()`` on rank 0 and ``Algorithm::execNonMaster()`` on all other ranks.
+
+To support running existing Python scripts without significant modification, and to be able to automatically determine execution modes based on input workspaces, workspaces with storage mode ``StorageMode::MasterOnly`` also exist on the non-master ranks.
+The default implementation of ``Algorithm::execNonMaster()`` creates an **uninitialized** (in the case of ``MatrixWorkspace``) workspace of the same type as the input workspace.
+If ``Algorithm::execNonMaster()`` is overridden, any workspaces that are created shall also be uninitialized and should have storage mode ``StorageMode::MasterOnly``.
+
+Given that the workspace on non-master ranks are not initialized, no methods of the workspace should be called, apart from ``Workspace::storageMode()``.
+Validators on the non-master ranks are thus also disabled.
+
+A typical implementation could look as follows:
+
+.. code-block:: c++
+
+  void MyAlg::execNonMaster() {
+    setProperty("OutputWorkspace", Kernel::make_unique<Workspace2D>(
+                                       Parallel::StorageMode::MasterOnly));
+  }
+
+
+Setting spectrum numbers
+------------------------
+
+Setting spectrum numbers via the legacy interface ``MatrixWorkspace::getSpectrum(size_t)::setSpectrumNo(specnum_t)`` is not supported in MPI runs and will throw an exception.
+The reason is that spectrum numbers are used to globally identify a spectrum and thus changing a spectrum number must be done globally, i.e., on all MPI ranks.
+Spectrum numbers should be set by using ``Indexing::IndexInfo`` and ``MatrixWorkspace::setIndexInfo()``, or rather by passing the ``IndexInfo`` to one of the workspace factory functions from ``DataObjects/WorkspaceCreation.h``.
+
+
+Workspace indices
+-----------------
+
+If a workspace is distributed, i.e., has storage mode ``StorageMode::Distributed`` workspaces indices lose their meaning.
+In particular, ``MatrixWorkspace::getNumberHistograms()`` will return the local number of spectra and not the global size of the workspace.
+For purposes of interaction with the user interface and for internal consistency a global equivalent of the 'workspace index' concept has been introduced.
+This index is represented by ``Indexing::GlobalSpectrumIndex``. [#spectrum-index]_
+
+The consequences are as follows:
+
+- Workspace indices should not be logged or written into output of other types such as tables. Instead spectrum numbers (``Indexing::SpectrumNumber``) or global spectrum indices (``Indexing::GlobalSpectrumIndex``) must be used.
+- The number of histograms in a workspace obtained from ``MatrixWorkspace::getNumberHistograms()`` may only be used for processing all spectra.
```

Comment:
"may only be used for processing all spectra.".. don't we need to qualify this by saying "may only be used for processing all spectra on the local rank"